### PR TITLE
Add a key to toggle map block bounds

### PR DESCRIPTION
**Pull request by grapereader**
It's often useful to know where the map block boundaries are for doing server admin work and the like. Also useful for developers.

Modded Minecraft had a similar visualization for chunks. Was it forge that added that?? 🤷‍♂️

So I've added a block bounds render to the HUD that just shows a wireframe cube around the current block the player is standing in. It hijacks the same material used to render the node selection. Also added a new key map, empty by default, to toggle it.

### CODE REVIEW

**Comment by AFCMS** [comment]
Nice feature! Could you add screenshots?

**Comment by Wuzzy2** [comment]
It's basically just a simple red outline for the whole mapblock, and that's it. It's strange that you apparently don't need the `debug` priv for it.

**Comment by voxel-minetest** [comment]
Such function would be very nice! 

> Modded Minecraft had a similar visualization for chunks. Was it forge that added that?? man_shrugging 

It is already implemented in the engine (at least since 1.14.4) by pressing F3(Debugging)+G 
;)

**Comment by Desour** [comment]
I would prefer to have such lines for all currently visible mapblocks, one block is not enough.

**Comment by grapereader** [comment]
Yep it's super simple and entirely client-side. Not sure how a privilege requirement could be enforced.

**Comment by voxel-minetest** [comment]
> I would prefer to have such lines for all currently visible mapblocks, one block is not enough.

I totally missed, that @grapereader just want the current mapblock to be shown. 
So I aggree with you and that's also how it is in Minecraft (to correct that).

**Comment by AFCMS** [comment]
> It's basically just a simple red outline for the whole mapblock, and that's it. It's strange that you apparently don't need the `debug` priv for it.

In mc, any player can show mapblock bound, so I am against using a priv

**Comment by grapereader** [comment]
screenshot_20210408_015151

It's easy enough to do the one block by re-using the selection box. I suppose we could shift that around and show all of them within some radius if we're okay with the extra draw calls.

**Comment by voxel-minetest** [comment]
> > It's basically just a simple red outline for the whole mapblock, and that's it. It's strange that you apparently don't need the `debug` priv for it.
> 
> In mc, any player can show mapblock bound, so I am against using a priv

Agree, cause you can't overreach other players with such function.

--

> Edit: Alright I was bored enough to do it. Here's what it looks like to show all blocks within a 5x5x5 cube.

Nice work! But why 5x5x5? A mapblock is 16x16x16.

> I'm not sure if it is more helpful [...]

Definitely

**Comment by appgurueu** [comment]
Note that this can be achieved without engine modification through the use of entities.

**Comment by grapereader** [comment]
> Nice work! But why 5x5x5? A mapblock is 16x16x16.

It starts to get cluttered, fast. It could probably be upped to 7x7x7 (that's 112x112x112 nodes). I tried 10x10x10 and it was like looking at the game through a screen door.

> Note that this can be achieved without engine modification through the use of entities.

It's funny I was just thinking last night that there must have been a reason it wasn't implemented, and that someone was going to point out it could be implemented with a mod. The only argument for it I can think of is that as a debug tool it is convenient to have in the engine before any mods. 🤷‍♂️ But I can go either way on it at this point.

**Comment by grapereader** [comment]
This can be done with a mod. Closing this.

**Comment by voxel-minetest** [comment]
> This can be done with a mod. Closing this.

But why? This should be in the engine...

**Comment by grapereader** [comment]
> But why? This should be in the engine...

Alright I'll re-open. Maintainers can make an executive decision.

**Comment by AFCMS** [comment]
> This can be done with a mod. Closing this.

How??

File: src/client/hud.cpp [hunk 1]
```diff
@@ -862,6 +862,58 @@ void Hud::drawSelectionMesh()
...
+void Hud::drawBlockBounds()
+{
+	if (!m_block_bounds_mode) {
+		return;	
+	}
+
+	video::SMaterial old_material = driver->getMaterial2D();
+	driver->setMaterial(m_selection_material);
+
+	auto pos = player->getStandingNodePos();
```
Comment (SmallJoker) [inline src/client/hud.cpp:883]: This is not a long type, and statically reading the code is easier with specified types.
```suggestion
	v3s16 pos = player->getStandingNodePos();
```

File: src/client/hud.cpp [hunk 2]
```diff
@@ -862,6 +862,58 @@ void Hud::drawSelectionMesh()
...
+
+	auto blockPos = v3s16(
+		floorf((float) pos.X / 16.0f),
+		floorf((float) pos.Y / 16.0f),
+		floorf((float) pos.Z / 16.0f)
+	);
```
Comment (SmallJoker) [inline src/client/hud.cpp:889]:
```suggestion
	v3s16 blockPos(
		(float)pos.X / MAP_BLOCKSIZE,
		(float)pos.Y / MAP_BLOCKSIZE,
		(float)pos.Z / MAP_BLOCKSIZE
	);
```
1. Use globally valid blocksize macro
2. Replace assignment initialization
3. Floating point truncation is equal to `floorf` and happens automatically for implicit integer conversions

File: src/client/hud.cpp [hunk 3]
```diff
@@ -862,6 +862,58 @@ void Hud::drawSelectionMesh()
...
+
+	auto offset = intToFloat(client->getCamera()->getOffset(), BS);
+
+	s8 radius = 2;
+
+	auto halfNode = v3f(BS, BS, BS) / 2.0f;
+
+	for (s8 x = -radius; x <= radius; x++) {
+		for (s8 y = -radius; y <= radius; y++) {
+			for (s8 z = -radius; z <= radius; z++) {
```
Comment (SmallJoker) [inline src/client/hud.cpp:899]: If wanted, you may nest them as follows to reduce indent levels:
```suggestion
	for (s8 x = -radius; x <= radius; x++)
	for (s8 y = -radius; y <= radius; y++)
	for (s8 z = -radius; z <= radius; z++) {
```

(attention for closing brackets)

File: src/client/hud.cpp [hunk 4]
```diff
@@ -862,6 +862,58 @@ void Hud::drawSelectionMesh()
...
+				auto blockOffset = v3s16(x, y, z);
+
+				if (blockOffset.getLength() == 0 || m_block_bounds_mode > 1) {
```
Comment (SmallJoker) [inline src/client/hud.cpp:902]: `getLengthSQ()` is faster and does the same

File: src/client/hud.h [hunk 5]
```diff
@@ -125,6 +128,8 @@ class Hud
 
 	scene::SMeshBuffer m_rotation_mesh_buffer;
 
+	u8 m_block_bounds_mode = 0;
```
Comment (SmallJoker) [inline src/client/hud.h:131]: Please change this to an enum value for clarity.
```C++
enum BlockBoundMode {
	BBM_??,
	BBM_??,
	BBM_??
};
BlockBoundMode m_block_bounds_mode = ???;
```

File: src/client/hud.cpp [hunk 6]
```diff
@@ -862,6 +862,58 @@ void Hud::drawSelectionMesh()
...
+					auto box = aabb3f(
+						intToFloat((blockPos + blockOffset) * 16, BS) - offset - halfNode,
+						intToFloat(((blockPos + blockOffset) * 16) + 15, BS) - offset + halfNode
+					);
```
Comment (SmallJoker) [inline src/client/hud.cpp:906]: Please make use of `MAP_BLOCKSIZE` or `MAP_BLOCKSIZE-1` respectively.
You might also be interested in the function `getNodeBox()` (numeric.h)  which does a quite similar thing.

**Review by SmallJoker — COMMENTED** [review]
The concept good, but the code/implementation is a bit fragile.

**Comment by appgurueu** [comment]
> > This can be done with a mod. Closing this.
> 
> How??

Well, you can't add a custom key to do this, but you can display mapblocks using entities.

File: src/client/hud.cpp (hunk 2 above)
Comment (grapereader) [reply src/client/hud.cpp:889]: Could have sworn there was an edge case involving negative numbers that floorf handled differently than truncation. But maybe that doesn't apply here... Will test.

File: src/client/hud.h (hunk 5 above)
Comment (grapereader) [reply src/client/hud.h:131]: As long as it's okay to do `m_block_bounds_mode = static_cast<BlockBoundsMode>(m_block_bounds_mode + 1)` as a way to cycle through them...

File: src/client/hud.cpp (hunk 1 above)
Comment (grapereader) [reply src/client/hud.cpp]: Can you tell I write TypeScript for my day job? 😂

**Comment by voxel-minetest** [comment]
Maybe it would be good to also have chunks and sectors (if in use) to be shown. 
Press key to show mapblocks, press again to show mapchunks...

**Comment by appgurueu** [comment]
I don't think we need yet another key for this though particular feature though.

**Review by SmallJoker — APPROVED** [review]
Works as expected.

**Comment by SmallJoker** [comment]
Afterthought: Would you mind to display a status message, similar to all other "special" keys?

**Comment by voxel-minetest** [comment]
minecraft handles this with shortcuts: 
Pressing F3 opens or closes the debugging, but when you hold F3 you can press another key to run a specific function, as I wrote before for exmaple F3+G for showing/hiding chunk borders.

File: src/client/hud.cpp [hunk 7]
```diff
@@ -862,6 +862,56 @@ void Hud::drawSelectionMesh()
...
+	if (m_block_bounds_mode == BLOCK_BOUNDS_OFF) {
+		return;	
+	}
+
+	video::SMaterial old_material = driver->getMaterial2D();
+	driver->setMaterial(m_selection_material);
+
+	v3s16 pos = player->getStandingNodePos();
+
+	v3s16 blockPos(
```
Comment (sfan5) [inline src/client/hud.cpp:885]: `v3s16 blockPos = pos / MAP_BLOCKSIZE;` should work equally well
(int divisions behaves identically to flooring)

File: src/client/hud.cpp [hunk 8]
```diff
@@ -862,6 +862,56 @@ void Hud::drawSelectionMesh()
...
+	s8 radius = 2;
+
+	v3f halfNode = v3f(BS, BS, BS) / 2.0f;
+
+	for (s8 x = -radius; x <= radius; x++)
+	for (s8 y = -radius; y <= radius; y++)
+	for (s8 z = -radius; z <= radius; z++) {
+		v3s16 blockOffset(x, y, z);
+
+		if (blockOffset.getLengthSQ() == 0 || m_block_bounds_mode == BLOCK_BOUNDS_ALL) {
```
Comment (sfan5) [inline src/client/hud.cpp:902]: Even easier, if mode isn't `BLOCK_BOUNDS_ALL` just set radius to 0.

File: src/client/render/core.cpp [hunk 9]
```diff
@@ -77,6 +77,7 @@ void RenderingCore::draw3D()
 	if (!show_hud)
 		return;
 	hud->drawSelectionMesh();
+	hud->drawBlockBounds();
```
Comment (sfan5) [inline src/client/render/core.cpp:80]: shouldn't the selection mesh be drawn over the block bounds?

File: src/client/hud.cpp (hunk 7 above)
Comment (SmallJoker) [reply src/client/hud.cpp:885]: Well yes they don't. https://github.com/minetest/minetest/pull/11172#discussion_r611200013

Comment (sfan5) [reply src/client/hud.cpp:885]: Ah of course, negative numbers exist.

File: src/client/hud.cpp (hunk 8 above)
Comment (grapereader) [reply src/client/hud.cpp:902]: This was actually my initial idea but I thought it would cause the loop to skip. Guess that was premature; the loop is <= not < 😁

**Comment by lhofhansl** [comment]
Tried it with MAP_BLOCKSIZE = 32. Works fine. +1

**Comment by sorcerykid** [comment]
> > This can be done with a mod. Closing this.
> 
> How??

I've long had a mapblock boundary display feature on the JT2 server. It uses an entity similar to Protector Redo, although the coloured grid is 16x16x16, of course. I released the mod several weeks ago:

https://github.com/sorcerykid/toolbox

image

I think it's nice to now have a basic means of visualizing mapblocks builtin to the engine, without installing a mod. Kudos!
